### DISCUSSION TITLE
How to well compute the gradient of a 'Material' object?

### DISCUSSION THREAD

**Discussion by lgab13** [idaholab/moose#18674] [Q&A General] [answered]
Hi moose experts，

I created my application but as I'm a beginner with MOOSE, I miss a lot of basic things. 

I defined a Real property and  I would like to define its gradient in order to use it in one of my PDE. How can I do this easily? 

Please consider the following basic example:

```C++
MyProperty::MyProperty(const InputParameters & parameters)
: Material(parameters),
  _Rmax(getParam<Real>("Rmax")),
  _profile(declareProperty<Real>("MyProfile"))
{
}

void
MyProperty::computeQpProperties()
{
  const auto rx = _q_point[_qp](0);
  const auto ry = _q_point[_qp](1);
  const auto rz = _q_point[_qp](2);
  const auto sqr_norm = rx*rx+ry*ry+rz*rz;
  _profile[_qp] = __Rmax*_Rmax - sqr_norm;
}
```
Here-above, I simply defined f(rx,ry,rz) and now, my goal is to work with  ' grad f ' in the weak form of a diffusive flux of my PDE. 

Is the following approach correct and if not, how do it?

```C++
MyProperty::MyProperty(const InputParameters & parameters)
: Material(parameters),
  _Rmax(getParam<Real>("Rmax")),
  _profile(declareProperty<Real>("MyProfile")),
  _grad_profile(declareProperty<RealGradient>("Mygradient"))
{
}

void
MyProperty::computeQpProperties()
{
  const auto rx = _q_point[_qp](0);
  const auto ry = _q_point[_qp](1);
  const auto rz = _q_point[_qp](2);
  const auto sqr_norm = rx*rx+ry*ry+rz*rz;
  _profile[_qp] = _Rmax*_Rmax - sqr_norm;
  _grad_profile[_qp][0] = - 2. * rx ;
  _grad_profile[_qp][1] = - 2. * ry ;
  _grad_profile[_qp][2] = - 2. * rz ;
}
```
In this example, the function clearly depends on the spatial coordinates. But, if the property only defines a constant per quadrature points, how can we built the gradient of this property in order to use it in the weak form of a diffusive flux ? When the gradient of the function f is defined as the pressure gradient in the NS equations, that's good for me. But, in case of diffusive flux, I didn't find examples. Maybe I have to populate an AuxVariable or to use an UserObject? 

best regards

lg

**Comment by GiudGiud** [comment, marked as answer]
Hello

Please consult https://github.com/idaholab/moose/discussions/18415 for more information on this topic. 

Guillaume

**Reply by lgab13** [reply]
Thank you  Guillaume. This is what I was looking for.  
best regards,
LG
